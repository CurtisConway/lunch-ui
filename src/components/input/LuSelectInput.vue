<template>
  <div :class="defaultClassList">
    <LuText
      tag="label"
      :for="computedId"
      size="subtitle"
      :class="labelClassList"
    >
      {{ label }}<span v-if="required">*</span>
    </LuText>
    <select
      ref="inputElement"
      :id="computedId"
      v-model="valueInterface"
      :multiple="multiple"
      @change="onChange"
      @focus="onFocus"
      @blur="onBlur"
    >
      <option
        v-for="(item, i) in mappedItems"
        :key="`${computedId}-option-${i}`"
        :value="item.value"
      >
        {{ item.label }}
      </option>
    </select>
    <LuTriangle :color="color" />
    <transition name="grow-fade">
      <LuErrorList :errors="errorList" v-if="!isValid && focus" />
    </transition>
  </div>
</template>

<script>
import colors from '../../assets/js/mixins/colors';
import borders from '../../assets/js/mixins/borders';
import spacing from '../../assets/js/mixins/spacing';
import events from '../../assets/js/mixins/events';
import inputs from './_inputs.mixin';
import validation from './_validation.mixin';
import LuTriangle from './partials/_LuTriangle.vue';

export default {
  name: 'LuSelectInput',
  mixins: [colors, borders, spacing, events, inputs, validation],
  components: {
    LuTriangle,
  },
  props: {
    items: {
      type: Array,
      default: () => [],
    },
    multiple: {
      type: Boolean,
      default: () => false,
    },
    value: {
      type: [String, Array],
      default: () => undefined,
    },
  },
  computed: {
    mappedItems() {
      return this.items.map((item, index) => {
        if (typeof item === 'string') {
          return {
            label: item,
            value: item,
          };
        }
        if (typeof item === 'object') {
          return {
            label: item.label || 'N/A',
            value: item.value || index,
          };
        }
        return {
          label: 'N/A',
          value: index,
        };
      });
    },
  },
};
</script>
