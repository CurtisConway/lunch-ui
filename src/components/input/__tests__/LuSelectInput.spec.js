import Vue from 'vue';
import Router from 'vue-router';

import LuSelectInput from '../LuSelectInput.vue';

import {
  createLocalVue,
  mount,
} from '@vue/test-utils';
import { compileToFunctions } from 'vue-template-compiler'

describe('LuSelectInput.vue', () => {
  let mountFunction;
  let router = Router;
  let localVue = typeof Vue;

  beforeEach(() => {
    router = new Router();
    localVue = createLocalVue();
    localVue.use(Router);

    mountFunction = (options = {}) => {
      return mount(LuSelectInput, {
        localVue,
        router,
        ...options,
      });
    }
  });

  it('should render component and match snapshot', () => {
    expect(mountFunction().html()).toMatchSnapshot();
  });

  it('should render component with items and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        items: ['test', 'test2'],
      },
    }).html()).toMatchSnapshot();
  });

  it('should render with a label prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        label: 'test',
      },
    }).html()).toMatchSnapshot();
  });

  it('should render with an inputStyle prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        inputStyle: 'underline',
      },
    }).html()).toMatchSnapshot();

    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        inputStyle: 'solid',
      },
    }).html()).toMatchSnapshot();

    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        inputStyle: 'solo',
      },
    }).html()).toMatchSnapshot();
  });

  it('should render with an errors list and match snapshot', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        errors: ['error one', 'error two'],
      },
    });
    wrapper.vm.$refs.textInput.$refs.inputElement.focus();
    await wrapper.vm.$nextTick();

    expect(wrapper.html()).toMatchSnapshot();
  });

  it('should render with a success prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        success: true,
      },
    }).html()).toMatchSnapshot();
  });

  it('should render with a persistentLabel prop and match snapshot', () => {
    expect(mountFunction({
      propsData: {
        elementId: 'testInput',
        persistentLabel: true,
      },
    }).html()).toMatchSnapshot();
  });

  it('should pass the value prop to the input element', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        items: ['test', 'test2'],
        value: 'test',
      },
    });

    expect(wrapper.vm.$refs.textInput.$refs.inputElement.value).toBe('test');
  });

  it('should emit a focus and blur event', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        items: ['test', 'test2'],
        value: 'test',
      },
    });

    const focus = jest.fn();
    const blur = jest.fn();
    const inputElement = wrapper.vm.$refs.textInput.$refs.inputElement;
    wrapper.vm.$on('focus', focus);
    wrapper.vm.$on('blur', blur);
    inputElement.focus();
    inputElement.blur();
    inputElement.focus();
    inputElement.blur();

    expect(focus.mock.calls).toHaveLength(2);
    expect(blur.mock.calls).toHaveLength(2);
  });

  it('should emit an input event', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        items: ['test', 'test2'],
        value: 'test',
      },
    });

    const event = jest.fn();
    wrapper.vm.$on('input', event);
    wrapper.vm.$refs.textInput.$refs.inputElement.focus();
    wrapper.vm.$refs.textInput.$refs.inputElement.value = 'test2';
    wrapper.vm.$refs.textInput.$refs.inputElement.blur();
    wrapper.vm.$refs.textInput.$refs.inputElement.focus();
    wrapper.vm.$refs.textInput.$refs.inputElement.value = 'test';
    wrapper.vm.$refs.textInput.$refs.inputElement.blur();

    expect(event.mock.calls).toHaveLength(2);
  });

  it('should be valid without any validation rules', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        items: ['test', 'test2'],
        rules: [],
      },
    });

    expect(wrapper.vm.isValid).toBe(true);
  });

  it('should not be valid with validation rules that fail', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        items: ['test', 'test2'],
        rules: [v => !!v || 'Required'],
      },
    });
    wrapper.vm.validateInput();
    expect(wrapper.vm.isValid).toBe(false);
  });

  it('should be valid with validation rules that pass', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        items: ['test', 'test2'],
        value: 'test',
        rules: [v => !!v || 'Required'],
      },
    });

    expect(wrapper.vm.isValid).toBe(true);
  });

  it('should be not valid after input if validation fails', async () => {
    const wrapper = mountFunction({
      propsData: {
        elementId: 'testInput',
        items: ['test', 'test2'],
        rules: [v => !!v || 'Required'],
      },
    });
    const textInput = wrapper.find('input');
    textInput.setValue('');
    wrapper.vm.validateInput();

    expect(wrapper.vm.isValid).toBe(false);
  });
});
