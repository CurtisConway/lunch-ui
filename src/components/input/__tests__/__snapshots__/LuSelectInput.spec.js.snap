// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`LuSelectInput.vue should render component and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-outline"><label class="lu-text text-subtitle text-left" for="input-1">
    <!----></label> <select id="input-1"></select>
  <!---->
</div>
`;

exports[`LuSelectInput.vue should render component with items and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-outline"><label class="lu-text text-subtitle text-left" for="input-5">
    <!----></label> <select id="input-5">
    <option value="test">
      test
    </option>
    <option value="test2">
      test2
    </option>
  </select>
  <!---->
</div>
`;

exports[`LuSelectInput.vue should render with a label prop and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-outline"><label class="lu-text text-subtitle text-left" for="input-9">
    test
    <!----></label> <select id="input-9"></select>
  <!---->
</div>
`;

exports[`LuSelectInput.vue should render with a persistentLabel prop and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-outline"><label class="lu-text text-subtitle text-left persistent" for="input-34">
    <!----></label> <select id="input-34"></select>
  <!---->
</div>
`;

exports[`LuSelectInput.vue should render with a success prop and match snapshot 1`] = `
<div class="lu-form-group text-green border-green lu-input-outline"><label class="lu-text text-subtitle text-left" for="input-30">
    <!----></label> <select id="input-30"></select>
  <!---->
</div>
`;

exports[`LuSelectInput.vue should render with an errors list and match snapshot 1`] = `
<div class="lu-form-group text-red border-red lu-input-outline"><label class="lu-text text-subtitle text-left focus" for="input-25">
    <!----></label> <select id="input-25"></select>
  <ul class="lu-error-list bg-red text-white pv-1 pr-2 pl-5 rad-1 grow-fade-enter grow-fade-enter-active">
    <li class="mv-1">
      error one
    </li>
    <li class="mv-1">
      error two
    </li>
  </ul>
</div>
`;

exports[`LuSelectInput.vue should render with an inputStyle prop and match snapshot 1`] = `
<div class="lu-form-group text-black border-black lu-input-underline"><label class="lu-text text-subtitle text-left" for="input-13">
    <!----></label> <select id="input-13"></select>
  <!---->
</div>
`;

exports[`LuSelectInput.vue should render with an inputStyle prop and match snapshot 2`] = `
<div class="lu-form-group text-black border-black lu-input-solid"><label class="lu-text text-subtitle text-left" for="input-17">
    <!----></label> <select id="input-17"></select>
  <!---->
</div>
`;

exports[`LuSelectInput.vue should render with an inputStyle prop and match snapshot 3`] = `
<div class="lu-form-group text-black border-black lu-input-solo elevation-3"><label class="lu-text text-subtitle text-left" for="input-21">
    <!----></label> <select id="input-21"></select>
  <!---->
</div>
`;
